mcnode/api.go

```diff
@@ -328,6 +327,81 @@ func (node *Node) httpDelete(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprintln(w, count)
 }
 
+// datastore interface
+type DataObject struct {
+	Data []byte `json:"data"`
+}
+
+// Get /data/get/{objectId}
+// Retrieves a data object from the datastore
+func (node *Node) httpGetData(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	key58 := vars["objectId"]
+	key, err := multihash.FromB58String(key58)
+	if err != nil {
+		apiError(w, http.StatusBadRequest, err)
+		return
+	}
+
+	data, err := node.ds.Get(Key(key))
+	if err != nil {
+		apiError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if data == nil {
+		apiError(w, http.StatusNotFound, UnknownObject)
+		return
+	}
+
+	dao := DataObject{data}
+	err = json.NewEncoder(w).Encode(dao)
+	if err != nil {
+		log.Printf("Error writing response body: %s", err.Error())
+	}
+}
+
+// POST /data/put
+// DATA: A stream of json-encoded data objects
+// Puts a batch of objects to the datastore
+// returns a stream of object ids (B58 encoded content multihashes)
+func (node *Node) httpPutData(w http.ResponseWriter, r *http.Request) {
+	rbody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("http/data/put: Error reading request body: %s", err.Error())
+		return
+	}
+
+	dec := json.NewDecoder(bytes.NewReader(rbody))
+	batch := make([][]byte, 0)
+
+	var dao DataObject
+loop:
+	for {
+		err = dec.Decode(&dao)
+		switch {
+		case err == io.EOF:
+			break loop
+		case err != nil:
+			apiError(w, http.StatusBadRequest, err)
+			return
+		default:
+			batch = append(batch, dao.Data)
```

Comment:
It is very unlikely that we have an OOM from this, the batch has to be very large. There is no batch size limit, but this is a user at fault if he makes a batch so large that it OOMs.

The error semantics are atomic write, either the entire batch is written to the database or an error results.

I think we really want to have a batch, so that we can amortize API call latency.